packages/markers/src/browser/markers-tree.ts

```diff
@@ -0,0 +1,115 @@
+/*
+* Copyright (C) 2017 TypeFox and others.
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+import { injectable, inject } from "inversify";
+import { Tree, ICompositeTreeNode, ITreeNode, ISelectableTreeNode, IExpandableTreeNode } from "@theia/core/lib/browser";
+import { MarkersManager } from './markers-manager';
+import { ProblemMarker } from './problem-marker';
+import { UriSelection } from "@theia/filesystem/lib/common";
+import { Diagnostic } from "vscode-languageserver-types";
+
+@injectable()
+export class MarkersTree extends Tree {
+
+    protected markerRoot: MarkerRootNode;
+
+    constructor( @inject(MarkersManager) protected readonly markersManager: MarkersManager) {
+        super();
+
+        markersManager.onDidChangeMarkers(() => this.refresh());
+
+        // TODO must be generalized
+        this.markerRoot = {
+            visible: false,
+            id: 'markerTree',
+            name: 'MarkerTree',
+            kind: 'problem',
+            children: [],
+            parent: undefined
+        };
+
+        this.root = this.markerRoot;
+    }
+
+    resolveChildren(parent: ICompositeTreeNode): Promise<ITreeNode[]> {
+        if (MarkerRootNode.is(parent)) {
+            return this.getFileNodes((parent as MarkerRootNode));
+        } else if (MarkerFileNode.is(parent)) {
+            return this.getMarkerNodes(parent);
+        }
+        return super.resolveChildren(parent);
+    }
+
+    getFileNodes(parent: MarkerRootNode): Promise<MarkerFileNode[]> {
+        const markerFiles = this.markersManager.getMarkerInformationByKind(this.markerRoot.kind);
+        const uriNodes: MarkerFileNode[] = [];
+        markerFiles.forEach(markerFile => {
+            const markerFileNode: MarkerFileNode = {
+                children: [],
+                expanded: false,
```

Comment:
We should preserve the expansion and selection state if there was already nodes for same ids